packages/adminrouter/extra/src/includes/server/master.conf

```diff
@@ -10,13 +10,42 @@ location /pkgpanda/active.buildinfo.full.json {
 
 # Group: Root
 # Description: DC/OS GUI
-root /opt/mesosphere/active/dcos-ui/usr;
+# root /opt/mesosphere/active/dcos-ui/usr;
```

Comment:
nginx is great at serving static files.

I guess I should see a design document about what the DC/OS UI service even is.

Reply:
@jgehrcke thanks for giving it a look!

The service is going to be serving UI static files. 

Otherwise the service would have to change files under admin router and that is, as to my knowledge, impossible to do as an atomic operation. Also that is less clean approach then having the service fully responsible for the files it manages.

Plus I'd like to get the reason behind the concern around not serving files with nginx? How often and how many users we have so that becomes risky serving them with a go server?

Design doc is there, I'm updating it https://docs.google.com/document/d/1HoOQTKFj1FulfHOM9JnPyMWKCoPKlv4kIatK7sn009k/edit#

Reply:
> A new component will fail. nginx is a proven work horse. When the UI update service fails it would be advisable if Admin Router could still serve the 'last known good' UI.

I bet we can configure what to do if `proxy_pass` fails do you have a config from the top of your head?

> Did we think about making the "UI update service" write a new directory to the file system and then make nginx atomically serve the new UI by updating a symlink to that directory?

We did consider this. That would mean we keep the knowledge about the version in two places: 
1. zookeeper
2. filesystem via symlink

I remember and most definitely sure switching a symlink to be _not_ an atomic operation. Thus changing the symlink under a running nginx in a distributed env doesn't appeal to us.

So the rough architecture would be:
1. the service serves a version of the UI
2. on the update request it downloads a new version and when all masters have it
3. switches to the new version

all the updates rollbacks would be switching one variable to another, no I/O and juggling with symlinks involved.

Reply:
How do you define the goal in terms of "atomically switching" here? Atomically within a single master node (so that the set of files that a single master node serves is exchanged atomically) or across all master nodes (so that all master nodes start serving the new set of files "at the same time")?

Reply:
Atomically within a single master node, there's no atomic in a distributed system, is there? :)